src/main/kotlin/com/empowerops/babel/backend.kt

```diff
@@ -11,13 +11,86 @@ import java.util.logging.Logger
+    fun BabelParser.ExprContext.asScalarExpression() = TypedExpression(ExpressionType.Scalar, this)
+
+    val exprType: Deque<TypedExpression> = LinkedList()
+
+    var problems: Set<ExpressionProblem> = emptySet()
         private set
 
-    override fun visitErrorNode(node: ErrorNode) {
-        problems += BabelExpressionProblem("syntax error", node.symbol.line, node.symbol.charPositionInLine)
+    override fun exitExpr(ctx: BabelParser.ExprContext) {
+
+        when {
+            ctx.childCount == 0 || ctx is ErrorNode -> {
+                // the most common case for this is when you having a dangling expression (eg 'x1 + x2 +')
+                // but it can happen in many cercomstances. In playing around I've found that
+                // a little bit of pollution here by adding an extra argument
+                // to the stack doesn't cause any new failure modes
+                // and allows analysis to continue on existing failure modes nicely.
+                // it does mean we cannot have a post-condition that this analysis is "clean"
+                // (eg a require(exprType.isEmpty()) after analysis)
+                exprType.push(ctx.asScalarExpression())
+                CodeGeneratingWalker.Log.fine { "pushed scalar type for imaginery node" }
+            }
+            ctx.callsLiteralOrVariable() -> exprType.push(ctx.asScalarExpression())
+            ctx.callsDynamicVariableAccess() -> exprType.push(ctx.asScalarExpression())
+            ctx.callsInlineExpression() -> {
+                val child = exprType.pop()
+                exprType.push(ctx.typedAs(child.expressionType))
+            }
+            ctx.callsAggregation() -> {
+                ensureIsScalar(exprType.pop())
+                ensureIsScalar(exprType.pop())
+                exprType.push(ctx.asScalarExpression())
+            }
+            ctx.callsBinaryFunction() -> {
+                ensureIsScalar(exprType.pop())
+                ensureIsScalar(exprType.pop())
+                exprType.push(ctx.asScalarExpression())
+            }
+            ctx.callsUnaryOpOrFunction() -> {
+                ensureIsScalar(exprType.pop())
+                exprType.push(ctx.asScalarExpression())
+            }
+            ctx.callsBinaryOp() -> {
+
+                ensureIsScalar(exprType.pop())
+                ensureIsScalar(exprType.pop())
+
+                when(ctx.getChild(1)){
+                    is BabelParser.GtContext, is BabelParser.GteqContext,
+                    is BabelParser.LtContext, is BabelParser.LteqContext -> {
+                        exprType.push(ctx.typedAs(ExpressionType.Boolean))
+                    }
+                    else -> exprType.push(ctx.asScalarExpression())
+                }
+            }
+            else -> TODO("unknown expr ${ctx.text}")
+        }
+    }
+
+    fun ensureIsScalar(expr: TypedExpression): Unit = when(expr.expressionType){
```

Comment:
`ensure` sounds a lot like it will throw an exception --when this function doesnt throw exceptions, it simply logs things.

change to sth like `reportIfNotScalar()`